python/hail/methods/misc.py

```diff
@@ -358,4 +369,185 @@ def window_by_locus(mt: MatrixTable, bp_window_size: int) -> MatrixTable:
+    >>> hl.array_windows(np.array([1, 2, 4, 4, 6, 8]), 2)
+    (array([0, 0, 1, 1, 2, 4]), array([2, 4, 5, 5, 6, 6]))
+
+    >>> hl.array_windows(np.array([-10.0, -2.5, 0.0, 0.0, 1.2, 2.3, 3.0]), 2.5)
+    (array([0, 1, 1, 1, 2, 2, 4]), array([1, 4, 6, 6, 7, 7, 7]))
+
+    Notes
+    -----
+    For a non-decreasing array ``a``, the resulting ``starts`` and ``stops``
+    arrays have the same length as ``a`` and the property that, for all indices
+    ``i``, ``[starts[i], stops[i])`` is the maximal range of indices ``j`` such
+    that ``a[i] - radius <= a[j] <= a[i] + radius``.
+
+    Index ranges are start-inclusive and stop-exclusive. This function is
+    especially useful in conjunction with
+    :meth:`.BlockMatrix.sparsify_row_intervals`.
+
+    Parameters
+    ----------
+    a: :obj:`ndarray` of signed integer or float values
+        Non-decreasing 1-dimensional array of values.
+    radius: :obj:`float`
+        Non-negative radius of window for values.
+
+    Returns
+    -------
+    (:class:`ndarray` of :obj:`int64`, :class:`ndarray` of :obj:`int64`)
+        Tuple of start indices array and stop indices array.
+    """
+    if radius < 0:
+        raise ValueError(f'array_windows: radius must be non-negative, found {radius}')
+    if a.ndim != 1:
+        raise ValueError("array_windows: 'a' must be 1-dimensional")
+    if not (np.issubdtype(a.dtype, np.signedinteger) or np.issubdtype(a.dtype, np.floating)):
+        if None in a:
+            raise ValueError(f"array_windows: 'a' contains a None value")
+        else:
+            raise ValueError(f"array_windows: 'a' must be an ndarray of signed integer or float values, "
+                             f"found dtype {str(a.dtype)}")
+    if (not np.all(a[:-1] <= a[1:])) or (a.size == 1 and np.isnan(a[0])):
+        raise ValueError("array_windows: 'a' must be non-decreasing with no nan elements")
+    size = a.size
+    if size > 0:
+        if a[0] - radius > a[0]:
+            raise ValueError('array_windows: underflow for a[0] - radius')
+        if a[-1] + radius < a[-1]:
+            raise ValueError('array_windows: overflow for a[-1] + radius')
+
+    starts, stops = np.zeros(size, dtype=np.int64),  np.zeros(size, dtype=np.int64)
+    j, k = 0, 0
+    for i in range(size):
+        min_val = a[i] - radius
+        while j < size and a[j] < min_val:
+            j += 1
+        starts[i] = j
+
+        max_val = a[i] + radius
+        while k < size and a[k] <= max_val:
+            k += 1
+        stops[i] = k
+
+    return starts, stops
+
+
+@typecheck(locus_table=Table,
+           radius=oneof(int, float),
+           value_expr=nullable(expr_float64))
+def locus_windows(locus_table, radius, value_expr=None):
+    """Returns start and stop indices for window around each locus.
+
+    Examples
+    --------
+    Windows with 2bp radius for one contig with positions 1, 2, 3, 4, 5:
+
+    >>> mt = hl.balding_nichols_model(n_populations=1, n_samples=5, n_variants=5)
```

Comment:
This might be dangerous to assign new matrix table to `mt`. I think each test is run independently, but I'm not 100% sure. The safest way is to add a new global variable to `conftest.py`. I think this is fine, but if you get doctest errors, that's why.